java/arcs/core/allocator/Arc.kt

```diff
@@ -0,0 +1,151 @@
+            }
+        }
+
+    private suspend fun onArcStateChange(handler: (ArcState) -> Unit) {
+        if (arcStateChangeHandlers.isEmpty()) {
+            registerChangeHandlerWithArcHosts()
+        }
+        arcStateChangeHandlers += handler
+        handler(arcState)
+    }
+
+    private suspend fun onArcStateChangeFiltered(stateToFilter: ArcState, handler: () -> Unit) {
+        if (arcState == stateToFilter) {
+            handler()
+        } else {
+            onArcStateChange {
+                if (it == stateToFilter) {
+                    handler()
+                }
+            }
+        }
+    }
+
+    /** Called whenever the [ArcState] changes to [Running]. */
+    suspend fun onRunning(handler: () -> Unit) = onArcStateChangeFiltered(Running, handler)
+
+    /** Called whenever the [ArcState] changes to [Stopped]. */
+    suspend fun onStopped(handler: () -> Unit) = onArcStateChangeFiltered(Stopped, handler)
+
+    /** Called whenever the [ArcState] changes to [Error]. */
+    suspend fun onError(handler: () -> Unit) = onArcStateChangeFiltered(Error, handler)
+
+    private fun fireArcStateChange() = arcStateChangeHandlers.forEach {
+        it(arcState)
+    }
+
+    private fun recomputeArcState() {
+        val states = arcStatesByHostId.values
+        arcState = when {
+            states.any { it == Deleted } -> Deleted
+            states.any { it == Error } -> Error
+            states.all { it == Running } -> Running
+            states.all { it == Stopped } -> Stopped
+            states.all { it == NeverStarted } -> NeverStarted
+            else -> Indeterminate
+        }
+    }
+
+    private suspend fun fetchCurrentStates() {
+        val partitionsFor = allocator.getPartitionsFor(arcId)
+        partitionsFor?.forEach {
+            val arcHost = allocator.lookupArcHost(it.arcHost)
+            arcStatesByHostId[it.arcHost] = arcHost.lookupArcHostStatus(it)
+        }
+    }
+
+    private suspend fun registerChangeHandlerWithArcHosts() {
+        val partitionsFor = allocator.getPartitionsFor(arcId)
+        // first poll the current states of all hosts
+        fetchCurrentStates()
+
+        // Register event listeners
+        partitionsFor?.forEach { partition ->
+            val arcHost = allocator.lookupArcHost(partition.arcHost)
+            arcHost.setOnArcStateChange(arcId) { _, state ->
+                arcStatesByHostId[partition.arcHost] = state
+                recomputeArcState()
+            }
+        }
+        recomputeArcState()
+    }
+
+    // suspend until a desired state is achieved
+    private suspend fun waitFor(state: ArcState): Arc {
+        if (arcState == state) {
+            return this
+        }
+
+        val deferred: CompletableDeferred<Arc> = CompletableDeferred()
+        onArcStateChangeFiltered(state) {
+            deferred.complete(this@Arc)
+        }
+
+        fetchCurrentStates()
+        recomputeArcState()
+        return deferred.await()
+    }
```

Comment:
This could be done without awaiting on a deferred by using `suspendCancellableCoroutine<Arc>`.

```kotlin
    private suspend fun waitFor(state: ArcState): Arc {
        if (arcState == state) return this
        
        return suspendCancellableCoroutine<Arc> { cont ->
            onArcStateChangeFiltered(state) { cont.resume(this@Arc) }

            fetchCurrentStates()
            recomputeArcState()
        }
    }
```

Also note: `onArcStateChangeFiltered(state)` could get called more than once.. so it would be good to un-register the callback after it's been called.